File: mobile-app/lib/models/note.dart
class Note {
  final String? id;
  final String title;
  final String content;
  final String ownerId;
  final String? folderId;
  final List<SharedWith> sharedWith;
  final bool isPublic;
  final String? publicToken;
  final int version;
  final String? lastModifiedBy;
  final bool isDeleted;
  final DateTime? deletedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  Note({
    this.id,
    required this.title,
    required this.content,
    required this.ownerId,
    this.folderId,
    this.sharedWith = const [],
    this.isPublic = false,
    this.publicToken,
    this.version = 1,
    this.lastModifiedBy,
    this.isDeleted = false,
    this.deletedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Note.fromJson(Map<String, dynamic> json) {
    return Note(
      id: json['_id'] ?? json['id'],
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      ownerId: json['owner_id']?['_id'] ?? json['owner_id'] ?? '',
      folderId: json['folder_id']?['_id'] ?? json['folder_id'],
      sharedWith: (json['shared_with'] as List?)
          ?.map((e) => SharedWith.fromJson(e))
          .toList() ?? [],
      isPublic: json['is_public'] ?? false,
      publicToken: json['public_token'],
      version: json['version'] ?? 1,
      lastModifiedBy: json['last_modified_by']?['_id'] ?? json['last_modified_by'],
      isDeleted: json['is_deleted'] ?? false,
      deletedAt: json['deleted_at'] != null ? DateTime.parse(json['deleted_at']) : null,
      createdAt: DateTime.parse(json['created_at'] ?? json['createdAt']),
      updatedAt: DateTime.parse(json['updated_at'] ?? json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'content': content,
      'folder_id': folderId,
      'version': version,
    };
  }
}

class SharedWith {
  final String userId;
  final String permission; // 'read', 'write', 'admin'

  SharedWith({
    required this.userId,
    required this.permission,
  });

  factory SharedWith.fromJson(Map<String, dynamic> json) {
    return SharedWith(
      userId: json['user_id']?['_id'] ?? json['user_id'] ?? '',
      permission: json['permission'] ?? 'read',
    );
  }
}

File: backend/models/Note.js
/**
 * Modèle pour les notes collaboratives (style Google Docs)
 */
const mongoose = require('mongoose');

const noteSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
    trim: true,
    maxlength: 255,
  },
  content: {
    type: String,
    default: '',
  },
  owner_id: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  folder_id: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Folder',
    default: null,
    index: true,
  },
  shared_with: [{
    user_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
    },
    permission: {
      type: String,
      enum: ['read', 'write', 'admin'],
      default: 'read',
    },
    added_at: {
      type: Date,
      default: Date.now,
    },
  }],
  is_public: {
    type: Boolean,
    default: false,
  },
  public_token: {
    type: String,
    default: undefined, // undefined plutôt que null pour éviter les problèmes d'index unique
  },
  version: {
    type: Number,
    default: 1,
  },
  last_modified_by: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
  },
  is_deleted: {
    type: Boolean,
    default: false,
    index: true,
  },
  deleted_at: {
    type: Date,
    default: null,
  },
  created_at: {
    type: Date,
    default: Date.now,
    index: true,
  },
  updated_at: {
    type: Date,
    default: Date.now,
    index: true,
  },
}, {
  timestamps: true,
});

// Index pour les requêtes fréquentes
noteSchema.index({ owner_id: 1, is_deleted: 1, created_at: -1 });
noteSchema.index({ folder_id: 1, is_deleted: 1 });
noteSchema.index({ 'shared_with.user_id': 1, is_deleted: 1 });
// Index unique sparse pour public_token (permet plusieurs null) - créé manuellement via script
// noteSchema.index({ public_token: 1 }, { unique: true, sparse: true });

// Méthode pour vérifier les permissions
noteSchema.methods.hasPermission = function(userId, permission = 'read') {
  if (!userId) {
    return false;
  }

  const mongoose = require('mongoose');
  
  // Normaliser userId en ObjectId si nécessaire
  let userIdObj;
  try {
    if (mongoose.Types.ObjectId.isValid(userId)) {
      userIdObj = new mongoose.Types.ObjectId(userId);
    } else {
      userIdObj = userId;
    }
  } catch (e) {
    userIdObj = userId;
  }

  // Normaliser owner_id pour la comparaison
  let ownerId;
  if (this.owner_id?._id) {
    ownerId = this.owner_id._id;
  } else if (this.owner_id) {
    ownerId = this.owner_id;
  } else {
    return false;
  }

  // Convertir en string pour comparaison fiable
  const ownerIdStr = ownerId.toString();
  const userIdStr = userIdObj.toString();
  
  // Propriétaire a tous les droits
  if (ownerIdStr === userIdStr) {
    return true;
  }

  // Vérifier les permissions partagées
  const share = this.shared_with.find(s => {
    if (!s.user_id) return false;
    
    let shareUserId;
    if (s.user_id._id) {
      shareUserId = s.user_id._id;
    } else {
      shareUserId = s.user_id;
    }
    
    return shareUserId.toString() === userIdStr;
  });
  
  if (!share) {
    return false;
  }

  const permissionLevels = { read: 1, write: 2, admin: 3 };
  const requiredLevel = permissionLevels[permission] || 1;
  const userLevel = permissionLevels[share.permission] || 0;

  return userLevel >= requiredLevel;
};

const Note = mongoose.model('Note', noteSchema);

module.exports = Note;


